src/isolation.rs

```diff
@@ -0,0 +1,225 @@
+pub fn create_temp_dir() -> TempDir {
+	let dir = Builder::new()
+		.permissions(Permissions::from_mode(0o700))
+		.prefix(&Uuid::new_v4().to_string())
+		.suffix("-uhyve")
+		.tempdir()
+		.ok()
+		.unwrap_or_else(|| panic!("The temporary directory could not be created."));
+
+	let dir_permissions = dir.path().metadata().unwrap().permissions();
+	assert!(!dir_permissions.readonly());
+
+	dir
+}
+
+/// HashMap matching a path in the guest OS ([String]) a path in the host OS ([OsString]).
+///
+/// Using a list of parameters stored in a [Vec<String>], this function creates
+/// a HashMap that can match a path on the host operating system given a path on
+/// the guest operating system.
+///
+/// See [crate::hypercall::open] to see this in practice.
+pub struct UhyveFileMap {
+	files: HashMap<String, OsString>,
+}
+
+impl UhyveFileMap {
+	/// Creates a UhyveFileMap.
+	///
+	/// * `parameters` - A list of parameters with the format `./host_path.txt:guest.txt`
+	pub fn new(parameters: &Option<Vec<String>>) -> UhyveFileMap {
+		if let Some(parameters) = parameters {
+			UhyveFileMap {
+				files: parameters
+					.iter()
+					.map(String::as_str)
+					.map(Self::split_guest_and_host_path)
+					.map(|(guest_path, host_path)| {
+						(
+							guest_path,
+							fs::canonicalize(&host_path).map_or(host_path, PathBuf::into_os_string),
+						)
+					})
+					.collect(),
+			}
+		} else {
+			UhyveFileMap {
+				files: Default::default(),
+			}
+		}
+	}
+
+	/// Separates a string of the format "./host_dir/host_path.txt:guest_path.txt"
+	/// into a guest_path (String) and host_path (OsString) respectively.
+	///
+	/// `parameter` - A parameter of the format `./host_path.txt:guest.txt`.
+	fn split_guest_and_host_path(parameter: &str) -> (String, OsString) {
+		let mut partsiter = parameter.split(":");
+
+		// Mind the order.
+		// TODO: Do this work using clap.
+		let host_path = OsString::from(partsiter.next().unwrap());
+		let guest_path = partsiter.next().unwrap().to_owned();
+
+		(guest_path, host_path)
+	}
+
+	/// Returns the host_path on the host filesystem given a requested guest_path, if it exists.
+	///
+	/// This function will look up the requested file in the UhyveFileMap and return
+	/// the corresponding path. Used in [`Hyper`] Internally, this function converts
+	/// &OsString to OsString. Otherwise, we would borrow UhyveFileMap in
+	/// [crate::hypercall::open] as an immutable, when we may need a mutable borrow
+	/// at a later point.
+	///
+	/// `guest_path` - The guest path. The file that the kernel is trying to open.
+	pub fn get_host_path(&mut self, guest_path: &str) -> Option<OsString> {
+		self.files.get(guest_path).map(OsString::from)
+	}
+
+	pub fn append_file_and_return_cstring(
+		&mut self,
+		guest_path: &str,
+		host_path: OsString,
+	) -> CString {
+		// TODO: Do we need to canonicalize the host_path?
+		self.files
+			.insert(String::from(guest_path), host_path.to_owned());
+
+		CString::new(host_path.as_bytes()).unwrap()
```

Comment:
I think, if you swap these lines you can safe one copy of the bytes:
```suggestion
		let ret = CString::new(host_path.as_bytes()).unwrap();
		self.files
			.insert(String::from(guest_path), host_path);
		ret
```